[Spec Constants] Improved handling of invalid spec. constants

RossBrunton opened the pull request:
Two main changes to how `Kernel/ProgramSetSpecializationConstants` are handled:
* They may now output either `INVALID_VALUE` or the new `INVALID_SPEC_ID` when the provided list is invalid.
* The OpenCL and level 0 adapters now respond to `UR_DEVICE_INFO_KERNEL_SET_SPECIALIZATION_CONSTANTS` with `false` rather than erroring out. This fixes some tests that were incorrectly not being skipped.
* `urKernelSetSpecializationConstants` now "implemented" (as a function that returns `UNSUPPORTED_FEATURE` for opencl and cuda.

aarongreig commented on test/conformance/program/program_adapter_cuda.match:
```diff
@@ -22,3 +22,6 @@
 {{OPT}}urProgramGetInfoTest.InvalidNullHandleProgram/NVIDIA_CUDA_BACKEND___{{.*}}___UR_PROGRAM_INFO_KERNEL_NAMES
 {{OPT}}urProgramLinkTest.Success/NVIDIA_CUDA_BACKEND___{{.*}}_
 {{OPT}}urProgramSetSpecializationConstantsTest.Success/NVIDIA_CUDA_BACKEND___{{.*}}_
```
I don't think any of the `ProgramSetSpecializationConstants` tests should be `{{OPT}}` for cuda and hip, neither support the entry point and we haven't gotten around to adding a device query to allow the tests to skip so they always fail.

ldrumm commented:
@jchlanda is our spec constants expert. Jakub, please take over this review

jchlanda commented on source/adapters/cuda/kernel.cpp:
```diff
@@ -377,3 +377,8 @@ urKernelSetArgSampler(ur_kernel_handle_t hKernel, uint32_t argIndex,
   return Result;
 }
+
+UR_APIEXPORT ur_result_t UR_APICALL urKernelSetSpecializationConstants(
```
Would it be worth it to do the same for HIP?

RossBrunton replied:
HIP already does this.

jchlanda replied:
Thank you

RossBrunton commented on include/ur_api.h:
```diff
@@ -4535,6 +4536,11 @@ typedef struct ur_specialization_constant_info_t {
 ///     - ::UR_RESULT_ERROR_INVALID_SIZE
 ///         + `count == 0`
+///     - ::UR_RESULT_ERROR_INVALID_VALUE
+///         + Any size specified in a pSpecConstant entry does not match the specialization constant in the module.
```
Reworded it to "A pSpecConstant entry contains a size that does not match that of the specialization constant in the module."

jchlanda replied:
Thank you.

PietroGhg reviewed (approved):
Native CPU lgtm, thank you

kbenzie commented:
> @RossBrunton I think the HIP match files need updated on this.

Forgot you were off. I would make the changes to this PR myself but the checkbox to allow project maintainers to make changes appears to have been disabled.

RossBrunton commented:
Can someone from @oneapi-src/unified-runtime-level-zero-write have a quick look at this? The changes for level_zero are changing an optional function from `DIE_NO_IMPLEMENTATION` to `  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;`

nrspruit commented on source/adapters/level_zero/device.cpp:
```diff
@@ -861,6 +861,8 @@ UR_APIEXPORT ur_result_t UR_APICALL urDeviceGetInfo(
   case UR_DEVICE_INFO_IMAGE_SRGB:
     return ReturnValue(ur_bool_t{false});
+  case UR_DEVICE_INFO_KERNEL_SET_SPECIALIZATION_CONSTANTS:
+    return ReturnValue(false);
```
Shouldn't this be ur_bool_t{false}?

RossBrunton replied:
In the time between creating this MR and now, it looks as if this change has been added via 646cc9ce7 , so this MR no longer touches level zero. There's a lot of device infos that return `bool`s rather than `ur_bool_t` here, so it might be worth having a separate change that updates all of them at once.

omarahmed1111 commented:
@RossBrunton Please rebase this when possible, I removed the ready to merge label and please readd it when rebased.

RossBrunton commented:
> @RossBrunton Please rebase this when possible, I removed the ready to merge label and please readd it when rebased.

Updated. Sorry for the delay.
